packages/server/src/core/types.ts

```diff
@@ -46,15 +46,23 @@ export type inferSubscriptionOutput<
 export type inferProcedureClientError<TProcedure extends AnyProcedure> =
   inferProcedureParams<TProcedure>['_config']['errorShape'];
 
-export type GetInferenceHelpers<TRouter extends AnyRouter> = {
+type GetInferenceHelpers<
+  TType extends 'input' | 'output',
+  TRouter extends AnyRouter,
+> = {
   [TKey in keyof TRouter['_def']['record']]: TRouter['_def']['record'][TKey] extends infer TRouterOrProcedure
     ? TRouterOrProcedure extends AnyRouter
-      ? GetInferenceHelpers<TRouterOrProcedure>
+      ? GetInferenceHelpers<TType, TRouterOrProcedure>
       : TRouterOrProcedure extends AnyProcedure
-      ? {
-          input: inferProcedureInput<TRouterOrProcedure>;
-          output: inferProcedureOutput<TRouterOrProcedure>;
-        }
+      ? TType extends 'input'
+        ? inferProcedureInput<TRouterOrProcedure>
+        : inferProcedureOutput<TRouterOrProcedure>
       : never
     : never;
 };
+
+export type GetInputInferenceHelpers<TRouter extends AnyRouter> =
+  GetInferenceHelpers<'input', TRouter>;
+
+export type GetOutputInferenceHelpers<TRouter extends AnyRouter> =
+  GetInferenceHelpers<'output', TRouter>;
```

Comment:
Could this not be done in user-land

```ts
import { type GetInferenceHelper } from "@trpc/server";

export type RouterInputs = GetInferenceHelpers<"input", AppRouter>;
```

Doesnt really matter just an alternative way to keep number of import/exports down?

Reply:
It could go either way, to be honest, functionality is the same

Reply:
I prefer the separate imports so the `"input"` and `"output"` generic isn't user-facing. Also makes it easier to change that type later if we want to.